v2/blockstore/readwrite.go

```diff
@@ -170,14 +174,40 @@ func (b *ReadWrite) resumeWithRoots(roots []cid.Cid) error {
 		return fmt.Errorf("cannot resume on CAR file with version %v", version)
 	}
 
-	// Check if file is finalized.
-	// A file is finalized when it contains a valid CAR v2 header with non-zero v1 offset.
-	// If finalized, cannot resume from.
+	// Check if file was finalized by trying to read the CAR v2 header.
+	// We check because if finalized the CARv1 reader behaviour needs to be adjusted since
+	// EOF will not signify end of CAR v1 payload. i.e. index is most likely present.
 	var headerInFile carv2.Header
-	if _, err := headerInFile.ReadFrom(internalio.NewOffsetReadSeeker(b.f, carv2.PragmaSize)); err == nil {
-		// TODO: we technically can; this could be a feature to do if asked for.
-		if headerInFile.CarV1Offset != 0 {
-			return fmt.Errorf("cannot resume from a finalized file")
+	_, err = headerInFile.ReadFrom(internalio.NewOffsetReadSeeker(b.f, carv2.PragmaSize))
+
+	// If reading CARv2 header succeeded, and CARv1 offset in header is not zero then the file is
+	// most-likely finalized. Check padding and truncate the file to remove index.
+	// Otherwise, carry on reading the v1 payload at offset determined from b.header.
+	if err == nil && headerInFile.CarV1Offset != 0 {
+		if headerInFile.CarV1Offset != b.header.CarV1Offset {
+			// Assert that the padding on file matches the given WithCarV1Padding option.
+			gotPadding := headerInFile.CarV1Offset - carv2.PragmaSize - carv2.HeaderSize
+			wantPadding := b.header.CarV1Offset - carv2.PragmaSize - carv2.HeaderSize
+			return fmt.Errorf(
+				"cannot resume from file with mismatched CARv1 offset; "+
+					"`WithCarV1Padding` option must match the padding on file."+
+					"Expected padding value of %v but got %v", wantPadding, gotPadding,
+			)
+		} else if headerInFile.CarV1Size != 0 {
+			// If header in file contains the size of car v1, then the index is most likely present.
+			// Since we will need to re-generate the index, as the one in file is flattened, truncate
+			// the file so that the Readonly.backing has the right set of bytes to deal with.
+			// This effectively means resuming from a finalized file will wipe its index even if there
+			// are no blocks put unless the user calls finalize.
+			if err := b.f.Truncate(int64(headerInFile.CarV1Offset + headerInFile.CarV1Size)); err != nil {
+				return err
+			}
+		} else {
+			// If CARv1 size is zero, since CARv1 offset wasn't, then the CARv2 header was
+			// most-likely partially written. Since we write the header last in Finalize then the
+			// file most-likely contains the index and we cannot know where it starts, therefore
+			// can't resume.
+			return errors.New("corrupt CARv2 header; cannot resume from file")
```

Comment:
this is unfortunate that partial writes that close uncleanly are now going to be something we aren't comfortable resuming.

It seems more likely that this is a case where the index has not been written at all, right?
It may be worth doing the slower recovery:
* Scan through the blocks in the carV1
* For each, see if the length is valid with regards to the length of the file
* either re-hash each block and make sure the contents match the CID, or ideally build up the index while doing the scan of lengths, and then do that hashing backwards until you find a valid block.
* if there is content after the last valid block, and it parses as an index, great
* if there is at most 1 final invalid block, we crashed in the middle of writing it and can truncate to before that and resume.
* otherwise we have an actual corrupt car and should force a new full export.

Reply:
> It seems more likely that this is a case where the index has not been written at all, right?

No, I think it's the case that the header is actually corrupt, since I have changed the code to write V2 header with the correct CarV1 size on every put. The rationale for doing this is in the commit message. TLDR; `Put` on finalised files that then closed without finalisation means header looks complete but actually missing latest put blocks.

On the slow recovery items listed above, I think that's useful regardless in recovering from files with tailing corrupt blocks.

Reply:
doing a write to the header on every put is potentially going to be a much worse write pattern for disk - instead of a linear write that an HDD would be happy with, we'll be jumping back and forth in our access pattern, which is going to potentially result in significantly slower performance.

Reply:
Agreed... It did not feel good writing it. What alternative can you think of?

Reply:
could we keep the carv1size as "-1" (or umaxint if it's unsigned) to indicate "until end of file"?

Reply:
We could but the issue is resuming from a valid and finalised CARv2 file. That file will contain the index and a complete header.

I guess what we could do is:
* if resuming from a file with populated header (i.e. non-zero carv1 size and offset) then remove index by truncating the file and _wipe off the valid header_  so that further `put`s without finalisation forces future resumptions to read file to the end.

Reply:
@willscott capture the workflow you kindly listed above in an improvement issue for another PR here: https://github.com/ipld/go-car/issues/157 Thank you for pointing this out